### PR TITLE
DT-1897: 🔧 Upgrade to spring boot 2.4.5 and java 16

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -58,8 +58,8 @@ dependencies {
 
     compileOnly 'org.projectlombok:lombok:1.18.20'
 
-    runtime 'org.hsqldb:hsqldb:2.5.1'
-    runtime 'org.flywaydb:flyway-core:7.7.2'
+    runtimeOnly 'org.hsqldb:hsqldb:2.5.1'
```
Comment (petergphillips) [inline build.gradle:61]: `runtime` doesn't work in gradle 7

File: build.gradle [hunk 2]
```diff
@@ -103,6 +103,7 @@ task testWithSchemaNomis(type: Test) {
     useJUnitPlatform()
     minHeapSize = "128m"
     maxHeapSize = "2048m"
+    jvmArgs = ["--illegal-access=permit"]
```
Comment (petergphillips) [inline build.gradle:106]: need this for the old cucumber tests

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderTransactionHistoryResourceTest.java [hunk 3]
```diff
@@ -232,8 +232,8 @@ public void When_GetOffenderTransactionHistory_And_BadFromDateFormat_Then_ErrorR
             OFFENDER_NO);
 
         assertThat(response.getBody().getStatus().intValue()).isEqualTo(HTTP_BAD_REQ);
-        assertThat(response.getBody().getDeveloperMessage()).startsWith("Unable to parse date time value \"2019-30-17\"");
-        assertThat(response.getBody().getUserMessage()).startsWith("Unable to parse date time value \"2019-30-17\"");
+        assertThat(response.getBody().getDeveloperMessage()).startsWith("Invalid value for MonthOfYear (valid values 1 - 12): 30");
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderTransactionHistoryResourceTest.java:235]: this is a revert to 5.3.4 spring behaviour - bug was introduced in spring 5.3.5 and fixed in spring 5.3.6

**Review by mikehalmamoj — APPROVED** [review]

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**